### ISSUE TITLE
Background install of PyBluez fails

### ISSUE THREAD

**Issue by Timmeey** [pchri/bt-presence-adapter#1]
Hey, great Idea using the presence of devices a binary sensors.

I wanted to give it a try, but after adding a regular bluettoth device via the addon->configure, nothing happens, no device shows up, searching for new devices doesn't bring anything up.

This is what i found in the logs

````
18-05-12 10:30:56.820 bt-presence:     Running setup.py install for PyBluez: finished with status 'error'
2018-05-12 10:30:56.821 bt-presence:     Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-xqwe7niz-build/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-y6i2r2la-record/install-record.txt --single-version-externally-managed --compile --prefix= --home=/tmp/tmpr7gzb6z3:
2018-05-12 10:30:56.822 bt-presence:     running install
2018-05-12 10:30:56.823 bt-presence:     running build
2018-05-12 10:30:56.824 bt-presence:     running build_py
2018-05-12 10:30:56.825 bt-presence:     creating build
2018-05-12 10:30:56.825 bt-presence:     creating build/lib.linux-armv7l-3.5
2018-05-12 10:30:56.826 bt-presence:     creating build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.827 bt-presence:     copying bluetooth/osx.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.828 bt-presence:     copying bluetooth/ble.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.828 bt-presence:     copying bluetooth/bluez.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.829 bt-presence:     copying bluetooth/msbt.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.830 bt-presence:     copying bluetooth/btcommon.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.831 bt-presence:     copying bluetooth/widcomm.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.831 bt-presence:     copying bluetooth/__init__.py -> build/lib.linux-armv7l-3.5/bluetooth
2018-05-12 10:30:56.832 bt-presence:     Fixing build/lib.linux-armv7l-3.5/bluetooth/osx.py build/lib.linux-armv7l-3.5/bluetooth/ble.py build/lib.linux-armv7l-3.5/bluetooth/bluez.py build/lib.linux-armv7l-3.5/bluetooth/msbt.py build/lib.linux-armv7l-3.5/bluetooth/btcommon.py build/lib.linux-armv7l-3.5/bluetooth/widcomm.py build/lib.linux-armv7l-3.5/bluetooth/__init__.py
2018-05-12 10:30:56.833 bt-presence:     Skipping optional fixer: buffer
2018-05-12 10:30:56.834 bt-presence:     Skipping optional fixer: idioms
2018-05-12 10:30:56.834 bt-presence:     Skipping optional fixer: set_literal
2018-05-12 10:30:56.835 bt-presence:     Skipping optional fixer: ws_comma
2018-05-12 10:30:56.836 bt-presence:     Fixing build/lib.linux-armv7l-3.5/bluetooth/osx.py build/lib.linux-armv7l-3.5/bluetooth/ble.py build/lib.linux-armv7l-3.5/bluetooth/bluez.py build/lib.linux-armv7l-3.5/bluetooth/msbt.py build/lib.linux-armv7l-3.5/bluetooth/btcommon.py build/lib.linux-armv7l-3.5/bluetooth/widcomm.py build/lib.linux-armv7l-3.5/bluetooth/__init__.py
2018-05-12 10:30:56.841 bt-presence:     Skipping optional fixer: buffer
2018-05-12 10:30:56.842 bt-presence:     Skipping optional fixer: idioms
2018-05-12 10:30:56.842 bt-presence:     Skipping optional fixer: set_literal
2018-05-12 10:30:56.843 bt-presence:     Skipping optional fixer: ws_comma
2018-05-12 10:30:56.844 bt-presence:     running build_ext
2018-05-12 10:30:56.844 bt-presence:     building 'bluetooth._bluetooth' extension
2018-05-12 10:30:56.845 bt-presence:     creating build/temp.linux-armv7l-3.5
2018-05-12 10:30:56.846 bt-presence:     creating build/temp.linux-armv7l-3.5/bluez
2018-05-12 10:30:56.846 bt-presence:     arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fdebug-prefix-map=/build/python3.5-RUbMX3/python3.5-3.5.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I./port3 -I/usr/include/python3.5m -c bluez/btmodule.c -o build/temp.linux-armv7l-3.5/bluez/btmodule.o
2018-05-12 10:30:56.847 bt-presence:     In file included from bluez/btmodule.c:20:0:
2018-05-12 10:30:56.848 bt-presence:     bluez/btmodule.h:5:33: fatal error: bluetooth/bluetooth.h: No such file or directory
2018-05-12 10:30:56.849 bt-presence:      #include <bluetooth/bluetooth.h>
2018-05-12 10:30:56.849 bt-presence:                                      ^
2018-05-12 10:30:56.850 bt-presence:     compilation terminated.
2018-05-12 10:30:56.851 bt-presence:     error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1

```

**Comment by Timmeey** [comment]
One little 
`sudo apt-get install libbluetooth-dev` and then a `pip install pybluez` 
lead to this error
`ImportError: No module named 'gattlib'`
installing
`sudo pip install gattlib ` then showed
`No package 'glib-2.0' found`
(i'm omitting all the intermediate build dependencies that were missing, since we all don't have all day to read through this :-P)

a quick
`sudo apt-get install libglib2.0-dev libboost-thread-dev libboost-python-dev` 
then let me finally install `sudo pip install gattlib `
Which still didn't help. In the end i followed https://raspberrypi.stackexchange.com/a/83243
but insteald using python, i used python3 install.

now it works.

But this cannot be the correct way of installing this great addon :-P

Maybe this can help to ease the installation process

**Comment by pchri** [comment]
Hi Tim

Thanks for your comments, and sorry for the delayed answer. I have been away for a few days. 

As you can see I still call this version 0.0.0 and it hasn't been added to the official addon list - simply because it isn't done. And it is exactly the dependencies that need a little work. So I'm very glad you found my addon (I don't know how), and that you have seemingly found a solution.

I will fix these issues using your suggestions, but I am a little pressed for time right now.

**Comment by Timmeey** [comment]
No worries. This wasn't meant as a harsh critique. I just hope that my comment and explanation of steps taken helps to address this issue at some point. Because I can really see this plug in as being very very useful. I already use it to make sure all my switches and lights are off when I'm definitely not at home. How did I find it? Just searched github for interesting projects and add-ons related to mozilla ist Gateway. Keep up the good work. And thanks for the plug in 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by pchri** [comment]
@mrstegeman can you give me a hint where to look for information about adding binaries to an adapter release. For this adapter I need libbluetooth-dev - and others for BLE.

**Comment by Timmeey** [comment]
Unfortunately i have no idea about how to integrate that into the adapter. I guess that would be a good ticket for the main project or the example adapter project. "Improve adapter  documentation, how to add dependencies"

**Comment by mrstegeman** [comment]
@pchri Distributing system-wide binaries, such as `libbluetooth-dev`, is difficult. The proper solution would be to have them included in the gateway image. To do so, you should modify the following files:
1. [prepare-base.sh](https://github.com/mozilla-iot/gateway/blob/master/image/prepare-base.sh#L46)
2. [post-upgrade.sh](https://github.com/mozilla-iot/gateway/blob/master/tools/post-upgrade.sh)
3. [README.md](https://github.com/mozilla-iot/gateway/blob/master/README.md) (to be kind to other people)

If you could submit a PR to the gateway, we should have no issues with including a few extra packages.

**Comment by pchri** [comment]
@Timmeey I have created release 0.1.0 which contains the necessary binary to make Bluetooth work on a Raspberry Pi 3. Unfortunately no BLE.

I'll start looking at @mrstegeman's suggestion soon. Hopefully that will lead to a nicer solution in the future.

**Comment by pchri** [comment]
@Timmeey I am happy to tell you that this has been fixed in release 0.2.0 which works with gateway version 0.5.0.

Notice that when you install the adapter it takes (on my RPi3) about 3 minutes for the system to install the necessary Python dependencies.
